magma/syntax/inline_combinational.py

```diff
@@ -0,0 +1,95 @@
+        self.target_map = target_map
+        self.name_count = 0
+
+    def _gen_new_name(self):
+        new_name = f"{self.prefix}{self.name_count}"
+        self.name_count += 1
+        return new_name
+
+    def visit_AugAssign(self, node):
+        if not isinstance(node.op, ast.MatMult):
+            return
+        key = immutable(node.target)
+        if key not in self.target_map:
+            retval_name = self._gen_new_name()
+            self.target_map[key] = retval_name
+        target = ast.Name(self.target_map[key], ast.Store())
+        return ast.Assign([target], node.value)
+
+
+class _RewriteToCombinational(Pass):
+    """
+    Replace assignment targets with a temporary value (so it is handled
+    properly by SSA), store the mapping from temporary values so they can be
+    wired up later, "inputs" will just read from the enclosing scope
+    """
+    def __init__(self, target_map):
+        super().__init__()
+        self.target_map = target_map
+
+    def rewrite(
+        self, tree: ast.AST, env: SymbolTable, metadata: MutableMapping
+    ) -> PASS_ARGS_T:
+        # prefix used for temporaries
+        prefix = gen_free_prefix(tree, env)
+
+        tree = _ToCombinationalRewriter(prefix, self.target_map).visit(tree)
+
+        # Return augassign targets for wiring
+        elts = ()
+        for value in self.target_map.values():
+            elts += (ast.Name(value, ast.Load()),)
+        if len(elts) > 1:
+            retval = ast.Tuple(elts, ast.Load())
+        else:
+            retval = elts[0]
+        tree.body.append(ast.Return(retval))
+
+        return tree, env, metadata
+
+
+class inline_combinational(apply_ast_passes):
+    def __init__(self, pre_passes=[], post_passes=[],
+                 debug: bool = False,
+                 env: Optional[SymbolTable] = None,
+                 path: Optional[str] = None,
+                 file_name: Optional[str] = None
+                 ):
+        self.target_map = {}
+        passes = (pre_passes + [_RewriteToCombinational(self.target_map),
+                                ssa(strict=False), bool_to_bit(),
+                                if_to_phi(lambda s, t, f: s.ite(t, f))] +
+                  post_passes)
+        super().__init__(passes=passes, env=env, debug=debug, path=path,
+                         file_name=file_name)
+
+    def exec(self, etree, stree, env, metadata):
+        fn = super().exec(etree, stree, env, metadata)
+        result = fn()
+        if not isinstance(result, tuple):
+            result = (result,)
+        for key, value in zip(self.target_map.keys(), result):
+            # Eval origin target in env to do wiring, not sure if there's a way
+            # we can avoid having to do eval here (other option is to codegen
+            # the original augassign in a new tree, but that's effectively the
+            # same as evaling)
+            wire(eval(astor.to_source(mutable(key)).rstrip(), {}, env), value)
```

Comment:
Could you try to look the name up in the env?  If its a name look up the name in env, if its and attribute getattr on the recursive call.

Reply:
That would work, but in theory there could also be a function call? At that point, we're effectively parsing and evaluating the code, so perhaps just using eval of the source is fine?

Reply:
Yeah in the face of a function we do need to do an eval, but I like trying to avoid evals where possible.

Reply:
So I suppose eval is fine.  The algorithm I am defining is basically an eval anyway.